java/sage/io/RemoteSageFile.java

```diff
@@ -0,0 +1,489 @@
+    len = (int)Math.min((long)len, maxRemoteSize - remoteOffset);
+
+    try
+    {
+      outStream.write(("READ " + remoteOffset + " " + len + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      inStream.readFully(b, off, len);
+    }
+    catch (IOException e)
+    {
+      reconnect();
+      outStream.write(("READ " + remoteOffset + " " + len + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      inStream.readFully(b, off, len);
+    }
+
+    remoteOffset += len;
+    return len;
+  }
+
+  @Override
+  public void readFully(byte[] b) throws IOException
+  {
+    readFully(b, 0, b.length);
+  }
+
+  @Override
+  public void readFully(byte[] b, int off, int len) throws IOException
+  {
+    if (len == 0)
+      return;
+
+    int bytesRead = read(b, off, len);
+
+    while (len > 0)
+    {
+      if (bytesRead == -1)
+        throw new EOFException("End of file reached!");
+
+      len -= bytesRead;
+      off += bytesRead;
+
+      if (len <= 0)
+        break;
+
+      bytesRead = read(b, off, len);
+    }
+  }
+
+  @Override
+  public synchronized void write(int b) throws IOException
+  {
+    if (readonly)
+      throw new IOException("Remote file is read only.");
+
+    try
+    {
+      outStream.write(("WRITE " + remoteOffset + " 1\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      outStream.write(b);
+    }
+    catch (IOException e)
+    {
+      reconnect();
+      outStream.write(("WRITE " + remoteOffset + " 1\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      outStream.write(b);
+    }
+
+    remoteOffset += 1;
+  }
+
+  @Override
+  public void write(byte[] b) throws IOException
+  {
+    write(b, 0, b.length);
+  }
+
+  @Override
+  public synchronized void write(byte[] b, int off, int len) throws IOException
+  {
+    if (readonly)
+      throw new IOException("Remote file is read only.");
+
+    if (len == 0)
+      return;
+
+    try
+    {
+      outStream.write(("WRITE " + remoteOffset + " " + len + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      outStream.write(b, off, len);
+    }
+    catch (IOException e)
+    {
+      reconnect();
+      outStream.write(("WRITE " + remoteOffset + " " + len + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      outStream.write(b, off, len);
+    }
+
+    remoteOffset += len;
+  }
+
+  @Override
+  public void randomWrite(long pos, byte[] b, int off, int len) throws IOException
+  {
+    if (readonly)
+      throw new IOException("Remote file is read only.");
+
+    long oldPosition = remoteOffset;
+    remoteOffset = pos;
+    try
+    {
+      write(b, off, len);
+    }
+    finally
+    {
+      remoteOffset = oldPosition;
+    }
+  }
+
+  @Override
+  public synchronized long skip(long n) throws IOException
+  {
+    if (n <= 0)
+      return 0;
+
+    long pos = remoteOffset;
+    long seek = Math.min(pos + n, length());
```

Comment:
You should probably optimize it to avoid calling length() unless required. You have this check in a few places, so you may just want to make a private method that does it to prevent code duplication (i.e. isLengthGreater(long))